processing/src/main/java/io/druid/segment/incremental/OffheapIncrementalIndex.java

```diff
@@ -393,82 +528,155 @@ public OffheapDimDim(String dimension)
      *
      * @see io.druid.segment.incremental.IncrementalIndexStorageAdapter.EntryHolderValueMatcherFactory#makeValueMatcher(String, String)
      */
-    public String get(String str)
+    @Override
+    public String intern(String str)
     {
-      final WeakReference<String> cached = cache.get(str);
-      if (cached != null) {
-        final String value = cached.get();
-        if (value != null) {
-          return value;
+      synchronized (cache) {
+        final WeakReference<String> cached = cache.get(str);
+        if (cached != null) {
+          final String value = cached.get();
+          if (value != null) {
+            return value;
+          }
         }
+        cache.put(str, new WeakReference<String>(str));
+        return str;
       }
-      cache.put(str, new WeakReference(str));
-      return str;
     }
 
+    private static final int ERROR_CLOSED = -2;
+
+    @Override
     public int getId(String value)
     {
-      return falseIds.get(value);
+      if (isClosed.get()) {
+        return ERROR_CLOSED; // special error code
+      }
+      Integer id;
+      do {
+        id = falseIds.get(value);
+      } while (null == id || id < 0);
+      return id;
     }
 
+    @Override
     public String getValue(int id)
     {
-      return falseIdsReverse.get(id);
+      if (id < 0) {
+        throw new IllegalArgumentException("id must be greater than or equal to 0");
+      }
+      String retVal;
+      final long start = Instant.now().getMillis();
```

Comment:
No, I was originally tinkering with the joda utils encapsulators, but ultimately found that simply using long looked cleaner and was simpler, the joda instant is a by-product of that, will change to use java standard stuff to reduce library dependency.